src/win/thread.c

```diff
@@ -170,6 +170,85 @@ int uv_thread_create(uv_thread_t *tid, void (*entry)(void *arg), void *arg) {
 }
 
 
+int uv_thread_setaffinity(uv_thread_t* tid,
+                          char* cpumask,
+                          char* oldmask,
+                          size_t mask_size) {
+  int i;
+  HANDLE hproc;
+  DWORD_PTR procmask;
+  DWORD_PTR sysmask;
+  DWORD_PTR threadmask;
+  DWORD_PTR oldthreadmask;
+  int cpumasksize;
+
+  cpumasksize = uv_cpumask_size();
+  if (mask_size < (size_t)cpumasksize)
+    return -EINVAL;
+
+  hproc = GetCurrentProcess();
+  if (!GetProcessAffinityMask(hproc, &procmask, &sysmask))
+    return uv_translate_sys_error(GetLastError());
+
+  threadmask = 0;
+  for (i = 0; i < cpumasksize; i++) {
+    if (cpumask[i]) {
+      if (procmask & (1 << i))
+        threadmask |= 1 << i;
+      else
+        return UV_EINVAL;
+    }
+  }
+
+  oldthreadmask = SetThreadAffinityMask(*tid, threadmask);
+  if (!oldthreadmask)
+    return uv_translate_sys_error(GetLastError());
+
+  if (oldmask != NULL) {
+    for (i = 0; i < cpumasksize; i++)
+      oldmask[i] = (oldthreadmask >> i) & 1;
+  }
+
+  return 0;
+}
+
+
+int uv_thread_getaffinity(uv_thread_t* tid,
+                          char* cpumask,
+                          size_t mask_size) {
+  int i;
+  HANDLE hproc;
+  DWORD_PTR procmask;
+  DWORD_PTR sysmask;
+  DWORD_PTR threadmask;
+  int cpumasksize;
+
+  cpumasksize = uv_cpumask_size();
+  if (mask_size < (size_t)cpumasksize)
+    return -EINVAL;
+
+  hproc = GetCurrentProcess();
+  if (!GetProcessAffinityMask(hproc, &procmask, &sysmask))
+    return uv_translate_sys_error(GetLastError());
+
+  threadmask = SetThreadAffinityMask(*tid, procmask);
+  if (!threadmask)
+    return uv_translate_sys_error(GetLastError());
+  SetThreadAffinityMask(*tid, threadmask);
+
+  for (i = 0; i < cpumasksize; i++)
+    cpumask[i] = (threadmask >> i) & 1;
+
+  return 0;
+}
+
+
+int uv_thread_detach(uv_thread_t* tid) {
+  CloseHandle(*tid);
```

Comment:
Shouldn't we check the return value of `CloseHandle()`

Reply:
We don't do so anywhere else (although a couple places in this file call `abort()` if it fails).

Reply:
Don't you think it would make sense returning the error if it fails?

Reply:
Not really. The only failure I would expect to see is ERROR_INVALID_HANDLE, and that would seem to represent a critical mistake in resource management (essentially a double-free). I don't know how the user would be expected to respond to a failure on cleanup. Similarly, the most nearly related function (`uv_thread_join`) doesn't check for errors (although, as I mentioned above, no call to `CloseHandle` in libuv currently returns an error).